include/mcl_3dl/lidar_measurement_models/lidar_measurement_model_beam.h

```diff
@@ -0,0 +1,207 @@
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef MCL_3DL_LIDAR_MEASUREMENT_MODELS_LIDAR_MEASUREMENT_MODEL_BEAM_H
+#define MCL_3DL_LIDAR_MEASUREMENT_MODELS_LIDAR_MEASUREMENT_MODEL_BEAM_H
+
+#include <algorithm>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <ros/ros.h>
+
+#include <pcl/point_types.h>
+#include <pcl_ros/point_cloud.h>
+
+#include <mcl_3dl/pf.h>
+#include <mcl_3dl/point_cloud_random_sampler.h>
+#include <mcl_3dl/raycast.h>
+#include <mcl_3dl/vec3.h>
+
+namespace mcl_3dl
+{
+template <class STATE_TYPE, class POINT_TYPE>
+class LidarMeasurementModelBeam : public LidarMeasurementModelBase<STATE_TYPE, POINT_TYPE>
+{
+  // POINT_TYPE must have label field (currently using intensity field as a label)
+  static_assert(std::is_same<pcl::PointXYZI, POINT_TYPE>(), "Supported POINT_TYPE is PointXYZI");
+
+private:
+  size_t num_points_;
+  size_t num_points_default_;
+  size_t num_points_global_;
+  float clip_far_sq_;
+  float clip_near_sq_;
+  float clip_z_min_;
+  float clip_z_max_;
+  float beam_likelihood_min_;
+  float beam_likelihood_;
+  float sin_total_ref_;
+  float map_grid_min_;
+  float map_grid_max_;
+
+  PointCloudRandomSampler<POINT_TYPE> sampler_;
+
+public:
+  LidarMeasurementModelBeam(const float x, const float y, const float z)
+  {
+    map_grid_min_ = std::min(std::min(x, y), z);
+    map_grid_max_ = std::max(std::max(x, y), z);
+  }
+
+  void loadConfig(
+      const ros::NodeHandle &nh,
+      const std::string &name)
+  {
+    ros::NodeHandle pnh(nh, name);
+
+    int num_points, num_points_global;
+    pnh.param("num_points", num_points, 3);
+    pnh.param("num_points_global", num_points_global, 0);
+    num_points_default_ = num_points_ = num_points;
+    num_points_global_ = num_points_global;
+
+    double clip_near, clip_far;
+    pnh.param("clip_near", clip_near, 0.5);
+    pnh.param("clip_far", clip_far, 4.0);
+    clip_near_sq_ = clip_near * clip_near;
+    clip_far_sq_ = clip_far * clip_far;
+
+    double clip_z_min, clip_z_max;
+    pnh.param("clip_z_min", clip_z_min, -2.0);
+    pnh.param("clip_z_max", clip_z_max, 2.0);
+    clip_z_min_ = clip_z_min;
+    clip_z_max_ = clip_z_max;
+
+    double beam_likelihood_min;
+    pnh.param("beam_likelihood", beam_likelihood_min, 0.2);
+    beam_likelihood_min_ = beam_likelihood_min;
+    beam_likelihood_ = powf(beam_likelihood_min, 1.0 / static_cast<float>(num_points));
```

Comment:
use `1.f` to avoid implicit casting.

Reply:
it never affect performance since 1.0 is constant and `1.0 / value` will be compiled as a single SSE's inverse operator.
(and also this is not performance critical function.)
do you think still it's better to explicitly define this as float?

Reply:
well then, put 1.0. ;)